tests/mocha/shortcut_registry_test.js

```diff
@@ -0,0 +1,355 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+suite('Keyboard Shortcut Registry Test', function() {
+  setup(function() {
+    sharedTestSetup.call(this);
+    this.registry = new Blockly.ShortcutRegistry();
+  });
+  teardown(function() {
+    sharedTestTeardown.call(this);
+  });
+
+  suite('Registering', function() {
+    test('Registering a shortcut', function() {
+      var testShortcut = {'name': 'test_shortcut'};
+      this.registry.register(testShortcut, true);
+      var shortcut = this.registry.registry_['test_shortcut'];
+      chai.assert.equal(shortcut.name, 'test_shortcut');
+    });
+    test('Registers shortcut with same name', function() {
+      var registry = this.registry;
+      var testShortcut = {'name': 'test_shortcut'};
+
+      registry.registry_['test_shortcut'] = [testShortcut];
+
+      var shouldThrow = function() {
+        registry.register(testShortcut);
+      };
+      chai.assert.throws(
+          shouldThrow, Error,
+          'Shortcut with name "test_shortcut" already exists.');
+    });
+    test(
+        'Registers shortcut with same name opt_allowOverrides=true',
+        function() {
+          var registry = this.registry;
+          var testShortcut = {'name': 'test_shortcut'};
+          var otherShortcut = {
+            'name': 'test_shortcut',
+            'callback': function() {}
+          };
+
+          registry.registry_['test_shortcut'] = [testShortcut];
+
+          var shouldNotThrow = function() {
+            registry.register(otherShortcut, true);
+          };
+          chai.assert.doesNotThrow(shouldNotThrow);
+          chai.assert.exists(registry.registry_['test_shortcut'].callback);
+        });
+  });
+
+  suite('Unregistering', function() {
+    test('Unregistering a shortcut', function() {
+      var testShortcut = {'name': 'test_shortcut'};
+      this.registry.registry_['test'] = [testShortcut];
+      chai.assert.isOk(this.registry.registry_['test']);
+      this.registry.unregister('test', 'test_shortcut');
+      chai.assert.isUndefined(this.registry.registry_['test']);
+    });
+    test('Unregistering a nonexistent shortcut', function() {
+      var consoleStub = sinon.stub(console, 'warn');
+      chai.assert.isUndefined(this.registry.registry_['test']);
+
+      var registry = this.registry;
+      chai.assert.isFalse(registry.unregister('test', 'test_shortcut'));
+      sinon.assert.calledWith(consoleStub, 'Keyboard shortcut with name "test" not found.');
```

Comment:
Just a minor note, instead of `calledWith` you could use `calledOnceWithExactly` to make sure there aren't any unexpected warnings being hidden.

Reply:
done